[Image-Set]: Support Resolution and Type as optional arguments

rreno opened the pull request:
#### e3554c53fff0e1d9ed207af749cd1dbaf7853300
<pre>
[Image-Set]: Support Resolution and Type as optional arguments
<a href="https://bugs.webkit.org/show_bug.cgi?id=225185">https://bugs.webkit.org/show_bug.cgi?id=225185</a>
rdar://77598590

Reviewed by Tim Nguyen.

Make resolution an optional argument to image-set. Also add support for
the type optional argument.

`type` is a hint that the author can provide to help us filter
unsupported image types from the set. We won&apos;t consider an unsupported
MIME type if it is specified.

<a href="http://w3c.github.io/csswg-drafts/css-images-4/#image-set-notation">http://w3c.github.io/csswg-drafts/css-images-4/#image-set-notation</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-parsing-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub-expected.txt:

* Source/WebCore/css/CSSImageSetOptionValue.cpp:
(WebCore::CSSImageSetOptionValue::Type::cssText const):
(WebCore::CSSImageSetOptionValue::CSSImageSetOptionValue):
(WebCore::CSSImageSetOptionValue::create):
(WebCore::CSSImageSetOptionValue::setType):
* Source/WebCore/css/CSSImageSetOptionValue.h:

* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::createStyleImage const):

* Source/WebCore/css/CSSValueKeywords.in:

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::ImageSetTypeCSSPrimitiveValueKnownTokenTypeFunctionConsumer::consume):
(WebCore::CSSPropertyParserHelpers::consumeImageSetOption):
(WebCore::CSSPropertyParserHelpers::consumeImageSet):

* Source/WebCore/rendering/style/StyleCursorImage.cpp:
(WebCore::StyleCursorImage::selectBestFitImage):

* Source/WebCore/rendering/style/StyleImageSet.cpp:
(WebCore::StyleImageSet::computedStyleValue const):

* Source/WebCore/rendering/style/StyleMultiImage.cpp:
(WebCore::StyleMultiImage::load):
    Early return if the image set was empty due to no supported image
    types being specified. We still ASSERT if the image set is non-empty
    and the image isn&apos;t actually a StyleImage type.
* Source/WebCore/rendering/style/StyleMultiImage.h:
    Augumet ImageWithScale to also carry a MIME type string if one was
    specified by the author so we can accurately report the computed
    value.

Canonical link: <a href="https://commits.webkit.org/260796@main">https://commits.webkit.org/260796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a98d3381e532d75351b9e869337f2d050d07dbb5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109482 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18606 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42244 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/980 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118644 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113367 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20072 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9803 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101759 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115237 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98182 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43159 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29835 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84922 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11348 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31178 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12011 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8110 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17377 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50782 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7481 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13743 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4420,30 +4452,61 @@ static RefPtr<CSSValue> consumeGeneratedImage(CSSParserTokenRange& range, const
-            if (resolution->floatValue() <= 0)
+    // Optional resolution and type in any order.
+    for (size_t i = 0; i < 2 && !args.atEnd(); ++i) {
+        if (auto optionalArg = consumeMetaConsumer<ImageSetResolutionOrTypeConsumer>(args, { }, { }, { }, { }, { })) {
```
nit: no abbreviation, can you name this `optionalArgument` or `parameter`?

rreno replied:
I expanded this to `optionalArgument` and I changed the argument to the function from `args` to `range` to match most of the other consumer functions calling that parameter `range`.

rreno committed:
[Image-Set]: Support Resolution and Type as optional arguments
https://bugs.webkit.org/show_bug.cgi?id=225185
rdar://77598590

Reviewed by Tim Nguyen.

Make resolution an optional argument to image-set. Also add support for
the type optional argument.

`type` is a hint that the author can provide to help us filter
unsupported image types from the set. We won't consider an unsupported
MIME type if it is specified.

http://w3c.github.io/csswg-drafts/css-images-4/#image-set-notation

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-parsing-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub-expected.txt:

* Source/WebCore/css/CSSImageSetOptionValue.cpp:
(WebCore::CSSImageSetOptionValue::Type::cssText const):
(WebCore::CSSImageSetOptionValue::CSSImageSetOptionValue):
(WebCore::CSSImageSetOptionValue::create):
(WebCore::CSSImageSetOptionValue::setType):
* Source/WebCore/css/CSSImageSetOptionValue.h:

* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::createStyleImage const):

* Source/WebCore/css/CSSValueKeywords.in:

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::ImageSetTypeCSSPrimitiveValueKnownTokenTypeFunctionConsumer::consume):
(WebCore::CSSPropertyParserHelpers::consumeImageSetOption):
(WebCore::CSSPropertyParserHelpers::consumeImageSet):

* Source/WebCore/rendering/style/StyleCursorImage.cpp:
(WebCore::StyleCursorImage::selectBestFitImage):

* Source/WebCore/rendering/style/StyleImageSet.cpp:
(WebCore::StyleImageSet::computedStyleValue const):

* Source/WebCore/rendering/style/StyleMultiImage.cpp:
(WebCore::StyleMultiImage::load):
    Early return if the image set was empty due to no supported image
    types being specified. We still ASSERT if the image set is non-empty
    and the image isn't actually a StyleImage type.
* Source/WebCore/rendering/style/StyleMultiImage.h:
    Augumet ImageWithScale to also carry a MIME type string if one was
    specified by the author so we can accurately report the computed
    value.

Canonical link: https://commits.webkit.org/260796@main
